File: lab2/lab2.js
const fs = require('fs');
const path = require('path');

// Перечислитель для цветов
const Color = {
    RED: '\x1b[31m',
    GREEN: '\x1b[32m',
    YELLOW: '\x1b[33m',
    BLUE: '\x1b[34m',
    MAGENTA: '\x1b[35m',
    CYAN: '\x1b[36m',
    WHITE: '\x1b[37m',
    RESET: '\x1b[0m'
};

class Printer {
    /**
     * @param {Object} options - Опции принтера
     * @param {Color} options.color - Цвет текста
     * @param {[number, number]} options.position - Позиция [x, y]
     * @param {string} options.symbol - Символ для псевдошрифта
     * @param {string} options.fontFile - Путь к файлу с описанием шрифта
     */
    constructor({ color, position, symbol, fontFile }) {
        this.color = color || Color.WHITE;
        this.position = position || [0, 0];
        this.symbol = symbol || '*';
        
        // Получаем абсолютный путь к файлу шрифта
        this.fontPath = fontFile ? path.join(__dirname, fontFile) : null;
        this.fontData = this.fontPath ? this.loadFont(this.fontPath) : null;
    }

    /**
     * Загружает шрифт из файла
     * @param {string} fontFile - Путь к файлу с описанием шрифта
     * @returns {Object} Объект с описанием символов шрифта
     */
    loadFont(fontPath) {
        try {
            if (!fs.existsSync(fontPath)) {
                console.error(`Font file not found at: ${fontPath}`);
                return null;
            }
            const data = fs.readFileSync(fontPath, 'utf8');
            return JSON.parse(data);
        } catch (error) {
            console.error('Error loading font file:', error.message);
            return null;
        }
    }

    /**
     * Устанавливает позицию курсора
     * @param {number} x 
     * @param {number} y 
     */
    setPosition(x, y) {
        process.stdout.write(`\x1b[${y};${x}H`);
    }

    /**
     * Сбрасывает настройки цвета
     */
    resetColor() {
        process.stdout.write(Color.RESET);
    }

    /**
     * Выводит текст с использованием псевдошрифта
     * @param {string} text 
     */
    printWithFont(text) {
        if (!this.fontData) {
            this.print(text);
            return;
        }

        const lines = {};
        const upperText = text.toUpperCase();

        // Собираем все линии для каждого символа
        for (const char of upperText) {
            if (this.fontData[char]) {
                const charLines = this.fontData[char].split('\n');
                for (let i = 0; i < charLines.length; i++) {
                    if (!lines[i]) lines[i] = '';
                    lines[i] += charLines[i].replace(/\*/g, this.symbol) + ' ';
                }
            } else if (char === ' ') {
                // Обработка пробела
                for (let i = 0; i < 5; i++) {
                    if (!lines[i]) lines[i] = '';
                    lines[i] += '     '; // 5 пробелов для пробела
                }
            }
        }

        // Выводим построчно
        let [x, y] = this.position;
        for (const line in lines) {
            this.setPosition(x, y + parseInt(line));
            process.stdout.write(this.color + lines[line] + Color.RESET);
        }
    }

    /**
     * Выводит текст
     * @param {string} text 
     */
    print(text) {
        this.setPosition(...this.position);
        process.stdout.write(this.color + text + Color.RESET);
    }

    /**
     * Статический метод для вывода текста
     * @param {string} text - Текст для вывода
     * @param {Color} color - Цвет текста
     * @param {[number, number]} position - Позиция [x, y]
     * @param {string} symbol - Символ для псевдошрифта
     * @param {string} fontFile - Путь к файлу с описанием шрифта
     */
    static print(text, color = Color.WHITE, position = [0, 0], symbol = '*', fontFile = null) {
        const printer = new Printer({ color, position, symbol, fontFile });
        if (fontFile) {
            printer.printWithFont(text);
        } else {
            printer.print(text);
        }
    }

    /**
     * Восстанавливает оригинальные настройки консоли
     */
    dispose() {
        this.resetColor();
        // Не можем восстановить позицию, так как не отслеживаем изменения
    }
}

// Демонстрация работы
function demo() {
    // Очищаем консоль
    console.clear();

    // Статическое использование
    Printer.print("Hello static!", Color.GREEN, [5, 5]);
    Printer.print("Big text!", Color.CYAN, [10, 10], '#', 'font.json');

    // Использование с контекстом
    const printer = new Printer({ 
        color: Color.MAGENTA, 
        position: [15, 15], 
        symbol: '@', 
        fontFile: 'font.json' 
    });
    
    try {
        printer.printWithFont("HELLO");
        printer.printWithFont("WORLD");
    } finally {
        printer.dispose();
    }

    // После выхода из блока
    console.log("\nBack to normal console output");
}

demo();

File: lab2/lab2.py
import json
import os
from enum import Enum
from typing import Tuple, Dict, List

class Color(Enum):
    RED = "\033[31m"
    GREEN = "\033[32m"
    BLUE = "\033[34m"
    YELLOW = "\033[33m"
    CYAN = "\033[36m"
    MAGENTA = "\033[35m"
    RESET = "\033[0m"

FONT_FILE = "lab2/font.json"

if os.path.exists(FONT_FILE):
    with open(FONT_FILE, "r", encoding="utf-8") as file:
        FONT_DATA = json.load(file)
else:
    raise FileNotFoundError(FONT_FILE)


class Printer:
    def __init__(self, color: Color, position: Tuple[int, int] | None = None, 
                 symbol: str = "*", font_size: int = 1):
        self.color = color
        self.symbol = symbol
        self.font_size = font_size
        self.original_position = (0, 0)
        self.position = position or self.original_position

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        print(Color.RESET.value, end="")

    @staticmethod
    def print(text: str, color: Color, position: Tuple[int, int] | None = None, 
              symbol: str = "*", font_size: int = 1):
        Printer._render_text(text, color, position, symbol, font_size)

    def print_text(self, text: str):
        Printer._render_text(text, self.color, self.position, self.symbol, self.font_size)

    @staticmethod
    def _scale_pattern(pattern: List[str], scale: int) -> List[str]:
        if scale == 1:
            return pattern
        
        scaled_pattern = []
        for line in pattern:
            scaled_line = ""
            for char in line:
                scaled_line += char * scale
            for _ in range(scale):
                scaled_pattern.append(scaled_line)
        return scaled_pattern

    @staticmethod
    def _render_text(text: str, color: Color, position: Tuple[int, int] | None, 
                    symbol: str, font_size: int):
        position = position or (0, 0)
        coordinate_settings = f"\033[{position[1]};{position[0]}H"
        print(color.value, end="")

        base_height = FONT_DATA['height']
        lines = ["" for _ in range(base_height * font_size)]
        
        for char in text.upper():
            if char in FONT_DATA["symbols"]:
                char_pattern = FONT_DATA["symbols"][char].split("\n")
                scaled_pattern = Printer._scale_pattern(char_pattern, font_size)
                
                for i, line in enumerate(scaled_pattern):
                    if symbol:
                        line = line.replace("*", symbol)
                    lines[i] += line + (" " * font_size * 2)

        for line in lines:
            print(coordinate_settings, line)
        print(Color.RESET.value, end="")


if __name__ == "__main__":
    # Статический вызов с разными размерами шрифта
    Printer.print("HELLO", Color.GREEN, (5, 12), "●", 2)
    Printer.print("WORLD", Color.BLUE, (5, 25), "▲", 3)
    Printer.print("SPACE", Color.RED, (10, 40), "●", 1)
    
    # Использование контекстного менеджера с разными размерами
    with Printer(Color.CYAN, (10, 40), "♥", 1) as small_printer:
        small_printer.print_text("HELLO")
    
    with Printer(Color.MAGENTA, (20, 40), "♦", 2) as medium_printer:
        medium_printer.print_text("HELLO")
    
    with Printer(Color.YELLOW, (30, 40), "♣", 3) as large_printer:
        large_printer.print_text("HELLO")
    